### PR TITLE
TestEnv improvements

### PR DESCRIPTION
**Pull request by rspeicher**
- Simplify cleaning the temporary testing path in TestEnv
- Don't run `gitlab:shell:install` if it's already installed
- Run git commands quietly

### CODE REVIEW

**Comment by TeatroIO** [comment]
I've prepared a stage to preview changes. <a href='http://tsigo-rs-testenv.c68bfdc4cac4c32c41e1a922e73062ea.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://tsigo-rs-testenv.c68bfdc4cac4c32c41e1a922e73062ea.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

File: spec/support/test_env.rb [hunk 1]
```diff
@@ -48,19 +39,27 @@ def disable_mailer
...
+
+    Dir[tmp_test_path].each do |entry|
+      # Keep gitlab-shell and gitlab-test
+      FileUtils.rm_rf(entry) unless File.basename(entry) =~ /\Agitlab-(shell|test)\z/
+    end
   end
 
   def setup_gitlab_shell
-    `rake gitlab:shell:install`
+    unless File.directory?(Rails.root.join(*%w(tmp tests gitlab-shell)))
```
Comment (jvanbaarsen) [inline spec/support/test_env.rb]: What is the speed improvement on this? Can you post a benchmark?

File: spec/support/test_env.rb [hunk 2]
```diff
@@ -48,19 +39,27 @@ def disable_mailer
...
 
-  def enable_mailer
-    NotificationService.any_instance.unstub(:mailer)
+  # Clean /tmp/tests
+  def clean_test_path
+    tmp_test_path = Rails.root.join('tmp', 'tests', '**')
+
+    Dir[tmp_test_path].each do |entry|
+      # Keep gitlab-shell and gitlab-test
+      FileUtils.rm_rf(entry) unless File.basename(entry) =~ /\Agitlab-(shell|test)\z/
```
Comment (jvanbaarsen) [inline spec/support/test_env.rb]: Can you please fix this warning?

File: spec/support/test_env.rb [hunk 3]
```diff
@@ -81,7 +80,7 @@ def setup_factory_repo
     end
 
     # We must copy bare repositories because we will push to them.
-    system(*%W(git clone --bare #{factory_repo_path} #{factory_repo_path_bare}))
+    system(*%W(git clone -q --bare #{factory_repo_path} #{factory_repo_path_bare}))
```
Comment (jvanbaarsen) [inline spec/support/test_env.rb]: Would there be a way to make this line under 83 characters? (Im not sure there is, so if none, we will ignore this one)

File: spec/support/test_env.rb (hunk 1 above)
Comment (tsigo) [inline spec/support/test_env.rb]: `master`:

```
gitlab master % time DISABLE_SPRING=true rspec spec/factories_spec.rb
Cloning into bare repository '/Users/tsigo/Code/rails/gitlab-development-kit/gitlab/tmp/tests/gitlab-test_bare'...
done.
................................................

Finished in 18.78 seconds
48 examples, 0 failures

Randomized with seed 56784

DISABLE_SPRING=true rspec spec/factories_spec.rb  22.02s user 6.94s system 98% cpu 29.524 total
```

`rs-testenv`:
gitlab rs-testenv % time DISABLE_SPRING=true rspec spec/factories_spec.rb
................................................

Finished in 7.54 seconds
48 examples, 0 failures

Randomized with seed 54243

DISABLE_SPRING=true rspec spec/factories_spec.rb  14.29s user 3.77s system 97% cpu 18.459 total
```

Nothing major, but not minor either if you're running specs outside of guard as I do.

File: spec/support/test_env.rb (hunk 3 above)
Comment (tsigo) [inline spec/support/test_env.rb]: Didn't bother with this one. You could change the `system` call to backticks and remove the `%W()` declaration, but I thought I remembered a GitLab coding guideline about always using this syntax for external processes.

**Comment by jvanbaarsen** [comment]
@tsigo Well, 10 seconds is major :) Can you make sure its mergeable?

**Comment by tsigo** [comment]
> @tsigo Well, 10 seconds is major :) Can you make sure its mergeable?

Should be good now.

**Pull request merged**